chainerrl/agents/td3.py

```diff
@@ -0,0 +1,370 @@
+            action = self.burnin_action_func()
+        else:
+            greedy_action = self.select_greedy_action(obs)
+            action = self.explorer.select_action(self.t, lambda: greedy_action)
+        self.t += 1
+
+        if self.last_state is not None:
+            assert self.last_action is not None
+            # Add a transition to the replay buffer
+            self.replay_buffer.append(
+                state=self.last_state,
+                action=self.last_action,
+                reward=reward,
+                next_state=obs,
+                next_action=action,
+                is_state_terminal=False)
+
+        self.last_state = obs
+        self.last_action = action
+
+        self.replay_updater.update_if_necessary(self.t)
+
+        return self.last_action
+
+    def act(self, obs):
+        return self.select_greedy_action(obs)
+
+    def batch_select_greedy_action(self, batch_obs):
+        with chainer.using_config('train', False), chainer.no_backprop_mode():
+            batch_xs = self.batch_states(batch_obs, self.xp, self.phi)
+            batch_action = self.policy(batch_xs).sample().array
+        return list(cuda.to_cpu(batch_action))
+
+    def batch_act(self, batch_obs):
+        return self.batch_select_greedy_action(batch_obs)
+
+    def batch_act_and_train(self, batch_obs):
+        """Select a batch of actions for training.
+
+        Args:
+            batch_obs (Sequence of ~object): Observations.
+
+        Returns:
+            Sequence of ~object: Actions.
+        """
+
+        if (self.burnin_action_func is not None
+                and self.policy_optimizer.t == 0):
+            batch_action = [self.burnin_action_func()
+                            for _ in range(len(batch_obs))]
+        else:
+            batch_greedy_action = self.batch_select_greedy_action(batch_obs)
+            batch_action = [
+                self.explorer.select_action(
+                    self.t, lambda: batch_greedy_action[i])
+                for i in range(len(batch_greedy_action))]
+
+        self.batch_last_obs = list(batch_obs)
+        self.batch_last_action = list(batch_action)
+
+        return batch_action
+
+    def batch_observe_and_train(
+            self, batch_obs, batch_reward, batch_done, batch_reset):
+        for i in range(len(batch_obs)):
+            self.t += 1
+            if self.batch_last_obs[i] is not None:
+                assert self.batch_last_action[i] is not None
+                # Add a transition to the replay buffer
+                self.replay_buffer.append(
+                    state=self.batch_last_obs[i],
+                    action=self.batch_last_action[i],
+                    reward=batch_reward[i],
+                    next_state=batch_obs[i],
+                    next_action=None,
+                    is_state_terminal=batch_done[i],
+                )
+                if batch_reset[i] or batch_done[i]:
+                    self.batch_last_obs[i] = None
```

Comment:
Shouldn't we have the `env_id` (https://github.com/chainer/chainerrl/blob/master/chainerrl/agents/ddpg.py#L442) since we might have an N-step replay buffer?

Reply:
Fixed.